### PR TITLE
feat(web): feature/A2-1090-UTC-Timezone-alignment

### PR DESCRIPTION
**Pull request by kandcdev**
## Describe your changes
A2-1090

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: appeals/web/src/server/lib/dates.js [hunk 1]
```diff
@@ -1,196 +1,213 @@
...
 	}
 
-	const hourString = hour.padStart(2, '0');
-	const minuteString = minute.padStart(2, '0');
+	let displayTimeString;
 
-	return `${hourString}:${minuteString}`;
-};
+	try {
+		displayTimeString = formatInTimeZone(dateISOString, timeZone, `h:mmaaaaa'm'`);
```
Comment (beejjacobs) [inline appeals/web/src/server/lib/dates.js:91]: This appears to work (though there are no units tests for this function directly), but I think can just be `h:mmaaa`

File: appeals/web/src/server/lib/dates.js [hunk 2]
```diff
@@ -1,196 +1,213 @@
...
 /**
- * @param {Date | number | string | null | undefined} date
- * @param {{ condensed?: boolean }} options
  * @returns {string}
  */
-export function dateToDisplayDate(date, { condensed = false } = {}) {
-	if (typeof date === 'undefined' || date === null) {
-		return '';
-	}
+export function getTodaysISOString() {
```
Comment (beejjacobs) [inline appeals/web/src/server/lib/dates.js:102]: This appears to be specifically "midnight today"? Seems fine, but perhaps the jsdoc comment could mention that.

File: appeals/web/src/server/lib/dates.js [hunk 3]
```diff
@@ -1,196 +1,213 @@
-import { formatInTimeZone } from 'date-fns-tz';
+import { formatInTimeZone, zonedTimeToUtc } from 'date-fns-tz';
 import enGB from 'date-fns/locale/en-GB/index.js';
-import logger from '#lib/logger.js';
-import { isValid, isBefore, isAfter, startOfDay, parseISO } from 'date-fns';
+import { isValid, isBefore, isAfter } from 'date-fns';
+import { padNumberWithZero } from '#lib/string-utilities.js';
 
 export const timeZone = 'Europe/London';
```
Comment (beejjacobs) [inline appeals/web/src/server/lib/dates.js:6]: `packages/appeals/constants/dates.js` includes this constant now, could be imported.

File: appeals/web/src/server/lib/mappers/appeal/submappers/site-visit-date.mapper.js [hunk 4]
```diff
@@ -13,9 +12,9 @@ export const mapSiteVisitDate = ({ appealDetails, currentRoute }) => ({
 			value: {
 				html:
 					dateAndTimeFormatter(
-						dateToDisplayDate(appealDetails.siteVisit?.visitDate),
-						convert24hTo12hTimeStringFormat(appealDetails.siteVisit?.visitStartTime),
-						convert24hTo12hTimeStringFormat(appealDetails.siteVisit?.visitEndTime)
+						dateISOStringToDisplayDate(appealDetails.siteVisit?.visitDate),
+						dateISOStringToDisplayTime12hr(appealDetails.siteVisit?.visitStartTime),
+						dateISOStringToDisplayTime12hr(appealDetails.siteVisit?.visitEndTime)
```
Comment (beejjacobs) [inline appeals/web/src/server/lib/mappers/appeal/submappers/site-visit-date.mapper.js:17]: Again we can tackle separately, but might make sense for `dateAndTimeFormatter` to accept three dates, and convert for display in that function.

**Review by beejjacobs — COMMENTED** [review]
Generally looks good. A few suggestions to consider.

File: appeals/web/src/server/lib/dates.js (hunk 1 above)
Comment (kandcdev) [reply appeals/web/src/server/lib/dates.js:91]: after few tries i found this solution on stackoverflow and it appears to work exactly as recommended by GDS

Comment (beejjacobs) [reply appeals/web/src/server/lib/dates.js:91]: Might be worth a link? e.g. to https://github.com/date-fns/date-fns/issues/946#issuecomment-452766765

File: appeals/web/src/server/lib/dates.js (hunk 2 above)
Comment (kandcdev) [reply appeals/web/src/server/lib/dates.js:102]: done

File: appeals/web/src/server/lib/dates.js (hunk 3 above)
Comment (kandcdev) [reply appeals/web/src/server/lib/dates.js]: done

File: appeals/web/src/server/lib/dates.js (hunk 1 above)
Comment (kandcdev) [reply appeals/web/src/server/lib/dates.js]: updated to `h:mmaaa` . turns out - no difference

File: appeals/api/src/server/common/validators/date-parameter.js [hunk 5]
```diff
@@ -34,30 +33,37 @@ const validateDateParameter = ({
...
+				if (!isEqual(value, recalculatedDate)) {
 					throw new Error(ERROR_MUST_BE_BUSINESS_DAY);
 				}
 			}
 
 			return true;
 		})
 		.custom(customFn)
-		.customSanitizer(joinDateAndTime);
+		.customSanitizer((value) => value);
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/api/src/server/common/validators/date-parameter.js:66]: yep, makes sense

File: appeals/api/src/server/endpoints/appeal-timetables/__tests__/appeal-timetables.test.js [hunk 6]
```diff
@@ -68,8 +64,8 @@ describe('appeal timetables routes', () => {
 				lpaQuestionnaireDueDate: futureDate
 			};
 			const householdAppealResponseBody = {
-				issueDeterminationDate: futureDateAndTime,
-				lpaQuestionnaireDueDate: futureDateAndTime
+				issueDeterminationDate: new Date(futureDate),
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/api/src/server/endpoints/appeal-timetables/__tests__/appeal-timetables.test.js:67]: I think we may need to keep the string and the Date as string is the input, date the output...

File: appeals/api/src/server/endpoints/appeals/appeals.formatter.js [hunk 7]
```diff
@@ -62,12 +62,16 @@ const formatMyAppeals = (appeal, linkedAppeals) => ({
 	appealTimetable: appeal.appealTimetable
 		? {
 				appealTimetableId: appeal.appealTimetable.id,
-				lpaQuestionnaireDueDate: appeal.appealTimetable.lpaQuestionnaireDueDate || null,
-				caseResubmissionDueDate: appeal.appealTimetable.caseResubmissionDueDate || null,
+				lpaQuestionnaireDueDate:
+					appeal.appealTimetable.lpaQuestionnaireDueDate?.toISOString() || null,
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/api/src/server/endpoints/appeals/appeals.formatter.js:66]: will address in the near future, lots of mapping and formatting to be improved there...

File: appeals/e2e/cypress/support/login.js [hunk 8]
```diff
@@ -13,6 +13,7 @@ const locators = {
 const timeout = async (ms) => await new Promise((resolve) => setTimeout(resolve, ms));
 
 azureSignIn = async (config) => {
+	console.log(config);
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/e2e/cypress/support/login.js:16]: whoops

**Review by fabfab-kc — APPROVED** [review]
Merging the beast

File: appeals/api/src/server/endpoints/site-visits/site-visits.controller.js [hunk 9]
```diff
@@ -48,9 +48,11 @@ const postSiteVisit = async (req, res) => {
 	/** @type { CreateSiteVisitData } */
 	const siteVisitData = {
 		appealId: Number(appealId),
-		visitDate: visitDate,
-		visitEndTime,
-		visitStartTime,
+		visitDate: isNaN(new Date(visitDate).getTime()) ? undefined : new Date(visitDate),
```
[the remark this answers is not in the document]
Comment (fabfab-kc) [reply appeals/api/src/server/endpoints/site-visits/site-visits.controller.js:51]: Not implemented yet

**Review by fabfab-kc — APPROVED** [review]

**Review by fabfab-kc — APPROVED** [review]

**Pull request merged**